clang-tidy: Add more `performance-*` checks and related fixes

hebasto opened the pull request.

aureleoules commented:
Concept ACK

aureleoules commented on src/.clang-tidy:
```diff
@@ -5,24 +5,13 @@ bugprone-use-after-move,
-performance-move-const-arg,
-performance-unnecessary-copy-initialization,
+performance-*,
+-performance-no-int-to-ptr,
+-performance-unnecessary-value-param,
```
Is there a reason why they these two are not enabled?

hebasto replied:
> Is there a reason why they these two are not enabled?

The required fixes seem too invasive for this PR.

aureleoules reviewed (approved):
ACK 66b5bdd791aca024996db182cfdec4439895bec0

MarcoFalke commented on src/.clang-tidy:
```diff
@@ -7,6 +7,7 @@ modernize-use-default-member-init,
 modernize-use-nullptr,
 performance-faster-string-find,
 performance-for-range-copy,
+performance-inefficient-string-concatenation,
```
Not sure. This makes the code pretty verbose and it is unclear whether append or strprintf should be used. Also, while this check is disabled for non-loops. In our code it is only triggered on errors in loops.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26642#issuecomment-1341293638).

hebasto commented:
Updated 66b5bdd791aca024996db182cfdec4439895bec0 -> 0524c30fbb05686a8cf8949a4b05d4a3c9a4c118 ([pr26642.01](https://github.com/hebasto/bitcoin/commits/pr26642.01) -> [pr26642.02](https://github.com/hebasto/bitcoin/commits/pr26642.02)):

- addressed @MarcoFalke [comment](https://github.com/bitcoin/bitcoin/pull/26642#discussion_r1042345453)

MarcoFalke commented on src/.clang-tidy:
```diff
@@ -9,6 +9,7 @@ performance-faster-string-find,
 performance-for-range-copy,
 performance-inefficient-vector-operation,
 performance-move-const-arg,
+performance-no-automatic-move,
```
Maybe split this out to make review/testing easier. It may be related to https://github.com/bitcoin/bitcoin/pull/25429 and https://github.com/bitcoin/bitcoin/pull/26308

hebasto replied:
> Maybe split this out to make review/testing easier. It may be related to #25429 and #26308

Done in #26758.

hebasto commented:
Rebased ecb0262a5a45f545a18c93e42317e013b8740fba -> 1f1b3fdc1872ae30127de363fc8c72f1649626de ([pr26642.03](https://github.com/hebasto/bitcoin/commits/pr26642.03) -> [pr26642.04](https://github.com/hebasto/bitcoin/commits/pr26642.04)) on top of the merged #26758.

MarcoFalke commented:
CI fails. Also, the first commit could be split up?

hebasto commented:
> Also, the first commit could be split up?

Done in https://github.com/bitcoin/bitcoin/pull/26905.

hebasto commented:
Rebased 1f1b3fdc1872ae30127de363fc8c72f1649626de -> 38f8938868f8ebebf3353e137957358399704a4b ([pr26642.04](https://github.com/hebasto/bitcoin/commits/pr26642.04) -> [pr26642.05](https://github.com/hebasto/bitcoin/commits/pr26642.05)) on top of the merged bitcoin-core/gui#686.

> CI fails.

Should be green now :)

hebasto commented:
Rebased 9e81656e92c91d128db8a5fcaecb0c385bb012cb -> 3f4e43caefb53ea726af07fb0802512264f82ae3 ([pr26642.06](https://github.com/hebasto/bitcoin/commits/pr26642.06) -> [pr26642.07](https://github.com/hebasto/bitcoin/commits/pr26642.07)) on top of the merged #26705 and #26935.

@martinus

Mind reviewing changes in `bench/nanobench.h`?

martinus commented:
> Mind reviewing changes in bench/nanobench.h?

ACK the nanobench.h changes. I'd like to keep the version in sync with the upstream repository at https://github.com/martinus/nanobench

So I'll have a look at the clang-tidy checks there, do a bunch of updates, and then create a PR to update the nanobench version here to stay in sync. That might take a while though.

martinus commented:
I created #27030 which compiles without warnings with the `.clang-tidy` changes here

hebasto commented:
> I created #27030 which compiles without warnings with the `.clang-tidy` changes here

Converting this PR to a draft for now.

hebasto commented:
> > I created #27030 which compiles without warnings with the `.clang-tidy` changes here
> 
> Converting this PR to a draft for now.

Rebased on the merged #27030, and is ready for a (final?) review :)

martinus commented on src/qt/walletmodel.h:
```diff
@@ -123,8 +123,8 @@ class WalletModel : public QObject
         // Move operator and constructor transfer the context
-        UnlockContext(UnlockContext&& obj) { CopyFrom(std::move(obj)); }
-        UnlockContext& operator=(UnlockContext&& rhs) { CopyFrom(std::move(rhs)); return *this; }
+        UnlockContext(UnlockContext&& obj) noexcept { CopyFrom(std::move(obj)); }
```
I'm not sure this is correct. `CopyFrom` is not noexcept, so marking this as noexcept can be dangerous. 
CopyFrom seems to use the copy assignment operator with its `*this = rhs;` which too is not noexcept. I think it can currently be marked as noexcept (based on the members here), but this is dangerous; e.g. when you add any data that needs allocating, the `noexcept` would be a lie and bad_alloc exceptions would cause the program to abort

hebasto replied:
@martinus 

The commented code is no longer relevant since https://github.com/bitcoin-core/gui/pull/711.

Nevertheless, I've dropped `performance-noexcept-move-constructor` commit for a follow up as it requires more thorough reviewing for similar cases.

martinus replied:
Good call, I find the `noexcept` stuff so brittle it's really hard to use. E.g. my changes to nanobench.h too were not completely correct because there's no guarantee that `std::string` move assignment is noexcept, see here: https://github.com/martinus/nanobench/pull/87

hebasto commented:
Drafted. Please review https://github.com/bitcoin-core/gui/pull/711 first.

hebasto commented:
Ready for a (final?) review now :)

martinus reviewed (commented):
I ran clang-tidy on this PR, and found a few more warnings:

```
/home/martinus/git/github.com/martinus/bitcoin/src/rpc/server.cpp:88:9: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
        vCommands.push_back(make_pair(entry.second.front()->category + entry.first, entry.second.front()));
        ^

/home/martinus/git/github.com/martinus/bitcoin/src/rpc/server.cpp:516:39: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
    for (const auto& i : mapCommands) commandList.emplace_back(i.first);
                                      ^

/home/martinus/git/github.com/martinus/bitcoin/src/rpc/util.cpp:612:9: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
        ret.emplace_back(arg.m_names);
```

hebasto commented:
Updated 2cdb84ca4c96cef03946b14ec46a7720f3ece098 -> 7db868b49b60de371d01cda1f8e5faaa0e18f822 ([pr26642.09](https://github.com/hebasto/bitcoin/commits/pr26642.09) -> [pr26642.10](https://github.com/hebasto/bitcoin/commits/pr26642.10)):

- rebased
- addressed @martinus's https://github.com/bitcoin/bitcoin/pull/26642#discussion_r1108879800

---

@martinus [wrote](https://github.com/bitcoin/bitcoin/pull/26642#discussion_r1108879800):
> I ran clang-tidy on this PR, and found a few more warnings

Is there a way to reproduce them in the CI environment?

hebasto commented on src/test/scheduler_tests.cpp:
```diff
@@ -71,6 +71,7 @@ BOOST_AUTO_TEST_CASE(manythreads)
 
     // As soon as these are created they will start running and servicing the queue
     std::vector<std::thread> microThreads;
+    microThreads.reserve(5);
```
Thanks! [Fixed](https://github.com/bitcoin/bitcoin/pull/26642#issuecomment-1440196497).

hebasto commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -268,12 +268,13 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
         key[i].MakeNewKey(true);
         BOOST_CHECK(keystore.AddKey(key[i]));
     }
+    std::vector<CPubKey> keys;
```
> I wonder why clang-tidy doesn't see these

So do I :)

martinus commented:
I don't know much about how the CI works, maybe it's possible to install a newer clang version? I'm using 15.0.7

hebasto commented:
> I don't know much about how the CI works, maybe it's possible to install a newer clang version? I'm using 15.0.7

Ah, that makes sense.

Maybe keep this PR diff CI-driven? I assume that bumping tools versions in a CI task will force us to fix more warnings, not just those you have already [mentioned](https://github.com/bitcoin/bitcoin/pull/26642#pullrequestreview-1302031747).

sedited reviewed (approved):
ACK 7db868b49b60de371d01cda1f8e5faaa0e18f822

I reproduced all changes here by running `clang-tidy` locally. They all seem like a clear improvement to me.

hebasto commented:
> I don't know much about how the CI works, maybe it's possible to install a newer clang version? I'm using 15.0.7

Done in https://github.com/bitcoin/bitcoin/pull/27311.

Making this PR a draft for now.

hebasto commented:
Rebased on top of the merged #27311.

All @martinus's comments have been addressed.

fanquake commented:
Seeing lots of https://cirrus-ci.com/task/6739422046060544?logs=ci#L339:
```bash
/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/rpc/server.cpp:88:9: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
        vCommands.push_back(make_pair(entry.second.front()->category + entry.first, entry.second.front()));
        ^
/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/rpc/server.cpp:516:39: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-i617 warnings generated.
```

hebasto commented:
> Seeing lots of https://cirrus-ci.com/task/6739422046060544?logs=ci#L339:

Sorry. Pushed a wrong branch.

Should be OK now.

TheCharlatan commented:
re-ACK [09383b6](https://github.com/bitcoin/bitcoin/pull/26642/commits/09383b612216b63f2aec0f1fffc889989c66f212)

Locally, I'm also getting a bunch of (though still on clang-14):
```
error: loop variable is copied but only used as const reference; consider making it a const reference [performance-for-range-copy,-warnings-as-errors]
```
after running:
```
bear --config src/.bear-tidy-config -- make -j $(nproc)
cd ./src/ && run-clang-tidy  -j $(nproc)
```

martinus commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2501,6 +2502,7 @@ TransactionError DescriptorScriptPubKeyMan::FillPSBT(PartiallySignedTransaction&
         } else {
             // Maybe there are pubkeys listed that we can sign for
             std::vector<CPubKey> pubkeys;
+            pubkeys.reserve(input.hd_keypaths.size());
```
nit: maybe add `+ 2` to the reserve because when the next `if` is true 2 more will be added. Otherwise the next `emplace_back` will most likely cause a capacity change

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26642#issuecomment-1484197246).

martinus reviewed (commented):
code review ACK 09383b612216b63f2aec0f1fffc889989c66f212

hebasto committed:
clang-tidy: Add `performance-faster-string-find` check

https://clang.llvm.org/extra/clang-tidy/checks/performance/faster-string-find.html

hebasto committed:
clang-tidy: Add `performance-inefficient-vector-operation` check

https://clang.llvm.org/extra/clang-tidy/checks/performance/inefficient-vector-operation.html

hebasto committed:
clang-tidy: Add `performance-type-promotion-in-math-fn` check

https://clang.llvm.org/extra/clang-tidy/checks/performance/type-promotion-in-math-fn.html

hebasto committed:
clang-tidy: Exclude `performance-*` checks rather including them

hebasto commented:
Updated 09383b612216b63f2aec0f1fffc889989c66f212 -> 03ec5b6f9ca3af28c9ce25cf2393e28ae852d808 ([pr26642.12](https://github.com/hebasto/bitcoin/commits/pr26642.12) -> [pr26642.13](https://github.com/hebasto/bitcoin/commits/pr26642.13), [diff](https://github.com/hebasto/bitcoin/compare/pr26642.12..pr26642.13)):

- addressed @martinus's [comment](https://github.com/bitcoin/bitcoin/pull/26642#discussion_r1148594367)

martinus commented:
ACK 03ec5b6f9ca3af28c9ce25cf2393e28ae852d808

TheCharlatan commented:
re-ACK [03ec5b6](https://github.com/bitcoin/bitcoin/pull/26642/commits/03ec5b6f9ca3af28c9ce25cf2393e28ae852d808)
